src/mscorlib/src/System/IO/MemoryStream.cs

```diff
@@ -402,31 +402,42 @@ public override Task<int> ReadAsync(Byte[] buffer, int offset, int count, Cancel
             return _buffer[_position++];
         }
 
+        public override void CopyTo(Stream destination, int bufferSize)
+        {
+            // Since we did not originally override this method, validate the arguments
+            // the same way Stream does for back-compat.
+            ValidateCopyToArguments(destination, bufferSize);
 
-        public override Task CopyToAsync(Stream destination, Int32 bufferSize, CancellationToken cancellationToken) {
-
-            // This implementation offers beter performance compared to the base class version.
-
-            // The parameter checks must be in sync with the base version:
-            if (destination == null)
-                throw new ArgumentNullException("destination");
+            // If we have been inherited into a subclass, the following implementation could be incorrect
+            // since it does not call through to Read() or Write() which a subclass might have overriden.  
+            // To be safe we will only use this implementation in cases where we know it is safe to do so,
+            // and delegate to our base class (which will call into Read/Write) when we are not sure.
+            if (GetType() != typeof(MemoryStream))
+            {
+                base.CopyTo(destination, bufferSize);
+                return;
+            }
             
-            if (bufferSize <= 0)
-                throw new ArgumentOutOfRangeException("bufferSize", Environment.GetResourceString("ArgumentOutOfRange_NeedPosNum"));
+            int originalPosition = _position;
+            int remaining = _length - originalPosition;
```

Comment:
@stephentoub 

> we should not be duplicating a whole block of code between them.

There's really not much that's being "duplicated" in this case-- just saving the position, subtracting it from the length, and updating `_position`.

> The cost of invoking InternalEmulateRead is not going to be measurable. If it's doing work that's unnecessary, then that work can be removed.

I agree the extra checks/call involved will likely have minimal overhead. My point was that having `InternalEmulateRead` makes the code less readable (the reader has to jump to yet another method to see what it does), and- removing the extra checks- it does not do any non-trivial work, so it does not help with code sharing.